signer.js

```diff
@@ -1,32 +1,91 @@
 'use strict'
 
-const cookieSignature = require('cookie-signature')
+const crypto = require('crypto')
 
-module.exports = function (secret) {
+function signerFactory (secret, factoryAlgorithm = 'sha256') {
   const secrets = Array.isArray(secret) ? secret : [secret]
-  const [signingKey] = secrets
+
+  for (const secret of secrets) {
+    if (typeof secret !== 'string') {
+      throw new TypeError('Secret key must be a string.')
+    }
+  }
+
+  const signingKey = secrets[0]
+
+  // validate that the algorithm is supported by the node runtime
+  try {
+    crypto.createHmac(factoryAlgorithm, signingKey)
+  } catch (e) {
+    throw new TypeError(`Algorithm ${factoryAlgorithm} not supported.`)
+  }
 
   return {
-    sign (value) {
-      return cookieSignature.sign(value, signingKey)
+    sign (value, secret = signingKey, algorithm = factoryAlgorithm) {
+      if (typeof value !== 'string') {
+        throw new TypeError('Cookie value must be provided as a string.')
+      }
+      return value + '.' + crypto
+        .createHmac(algorithm, secret)
+        .update(value)
+        .digest('base64')
+        // remove base64 padding (=) as it has special meaning in cookies
+        .replace(/=+$/, '')
     },
-    unsign (signedValue) {
-      let valid = false
-      let renew = false
-      let value = null
-
-      for (const key of secrets) {
-        const result = cookieSignature.unsign(signedValue, key)
-
-        if (result !== false) {
-          valid = true
-          renew = key !== signingKey
-          value = result
-          break
+    unsign (signedValue, keys = secrets, algorithm = factoryAlgorithm) {
+      if (typeof signedValue !== 'string') {
+        throw new TypeError('Signed cookie string must be provided.')
+      }
+      const value = signedValue.slice(0, signedValue.lastIndexOf('.'))
+      const actual = Buffer.from(signedValue)
+
+      for (const key of keys) {
+        const expected = Buffer.from(this.sign(value, key, algorithm))
+        if (
+          expected.length === actual.length &&
+          crypto.timingSafeEqual(expected, actual)
+        ) {
+          return {
+            valid: true,
+            renew: key !== keys[0],
+            value
+          }
         }
       }
 
-      return { valid, renew, value }
+      return {
+        valid: false,
+        renew: false,
+        value: null
+      }
+    }
+  }
+}
+
+// create a signer-instance, with dummy secret
+const signer = signerFactory(['dummy'])
```

Comment:
This seems really unsafe.

Reply:
It is just to pass the secret validation. The secret itself will be validated again in the corresponding sign and unsign function